src/datasets/low_carbon_london/preprocess_lcl.py

```diff
@@ -0,0 +1,244 @@
+
+    Returns:
+        pd.DataFrame: Output dataframe
+    """
+    logger.info("🗑 Dropping dupes and filling nulls with 0")
+    df_out = df.copy()
+    df_out = df_out.sort_values(
+        by=["LCLid", "date", "settlement_period"], ascending=True
+    )
+    df_out = df_out.drop_duplicates(
+        subset=["LCLid", "date", "settlement_period"], keep="last"
+    )
+    df_out["kwh"] = df_out["kwh"].replace("Null", np.float64())
+    df_out["kwh"] = df_out["kwh"].astype(float)
+    return df_out
+
+
+def filter_missing_kwh(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Drop dates where we don't have full 48 readings
+    for a given LCLid and date.
+
+    Args:
+        df (pd.DataFrame): _description_
+
+    Returns:
+        pd.DataFrame: _description_
+    """
+    logger.info("🔍 Filtering missing kwh readings")
+    id_col = ["LCLid"]
+    merge_cols = id_col + ["date"]
+    df_group = df.groupby(merge_cols)[["kwh"]].count().reset_index()
+    df_group["required_len"] = 48  # 48 hh readings
+
+    df_full_data = df_group.query("required_len==kwh")  # Has all required data
+    df_out = df_full_data[merge_cols].merge(df, on=merge_cols, how="inner")
+    return df_out
+
+
+def pack_smart_meter_data_into_arrays(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Pack smart meter data into 48-dimensional arrays
+
+    Args:
+        df (pd.DataFrame): Input dataframe
+
+    Returns:
+        pd.DataFrame: Output dataframe
+    """
+    logger.info("👝 Packing time series into arrays")
+    df_out = df.copy()
+    df_out = df_out.sort_values(
+        by=["LCLid", "date", "settlement_period"], ascending=True
+    )
+    groupby_cols = ["LCLid", "stdorToU", "month_end", "month"]
+    groupby_cols = groupby_cols + ["dayofweek", "day", "date"]
+
+    df_out = pd.DataFrame(
+        df_out.groupby(groupby_cols)["kwh"]
+        .agg(lambda x: x.tolist())
+        .reset_index()
+    )
+    return df_out
+
+
+def get_mean_and_std(df: pd.DataFrame) -> Tuple[float, float]:
+    """
+    Calculate the mean and standard deviation of the dataset
+
+    Args:
+        df (pd.DataFrame): Input dataframe
+
+    Returns:
+        Tuple[float, float]: Mean and standard deviation
+    """
+    mean = np.mean(df["kwh"])
+    std = np.std(df["kwh"])
+    return mean, std
+
+
+def preprocess_pipeline(file_path: str, out_path: str):
+
+    df = load_data(file_path)
+    df = extract_date_features(df)
+    df = parse_settlement_period(df)
+    df = drop_dupes_and_replace_nulls(df)
+    df = filter_missing_kwh(df)
+
+    mean, stdev = get_mean_and_std(df)
+    df = pack_smart_meter_data_into_arrays(df)
+
+    os.makedirs(out_path, exist_ok=True)
+    df.to_csv(f"{out_path}/lcl_data.csv", index=False)
+
+    mean_std_dict = {"mean": mean, "stdev": stdev}
+    with open(f"{out_path}/mean_std.csv", "w") as f:
+        w = csv.DictWriter(f, mean_std_dict.keys())
+        w.writeheader()
+        w.writerow(mean_std_dict)
+
+
+def preprocess_lcl_data():
```

Comment:
❓ could this be a bit confusing?  i.e. if you forget to do `split=True` you will still get 4 different datasets which are all copies of the same dataset but named as if they have already been split...

Reply:
So if you forget to split:

- If you have never split your dataset before, you'll get a `FileNotExist` error, as this function has no data to load.
- If you have already split your dataset, and you want to run preprocess code again for some reason, you don't need to split it again and could just run with `split=False`.

The intention is to allow users to avoid having to rerun any bits of preprocessing code as they all take time to run.

Reply:
Yes sorry you're right, I'm not sure what I was thinking at the time!  maybe I was starting to get ill 😆